File: 06b-LambdaCalculus_bak/_code/_haskell/03-ycombinator/ycombinator.hs
-- from: https://gist.github.com/JustinSDK/9c6c82775f422cc87a73
import Unsafe.Coerce

y :: (a -> a) -> a
y = \f -> (\x -> f (unsafeCoerce x x))(\x -> f (unsafeCoerce x x))

-- Ref: [Y Combinator in Haskell](http://stackoverflow.com/questions/4273413/y-combinator-in-haskell)
-- a fibonacci example
main = putStrLn $ show $ y (\fact -> \n -> if n < 2 then 1 else n * fact (n - 1)) 5

File: 04b-λCalculus書/_more/_book_bak/函數式編程/07-combinator/02-ycombinator/ChatGPT/ycombinator.py
# 定義 Y-Combinator
def Y(f):
    return (lambda x: f(lambda v: x(x)(v)))(lambda x: f(lambda v: x(x)(v)))

# 使用 Y-Combinator 實現階乘函數
factorial = Y(lambda f: lambda n: 1 if n == 0 else n * f(n - 1))

# 測試
print(factorial(5))  # 輸出 120
